### ISSUE TITLE
rpc: the arguments for api method are incorrect

### ISSUE THREAD

**Issue by Snikko** [paritytech/substrate#14728]
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Description of bug

I want to add rpc function in my substrate blockchain to supply restful apis for others.But some mistake occur when the project is built.The errors show like this: 
--------------------------------------------------------------------------------------
45  |         api.get_value(&at,v).map_err(runtime_error_into_rpc_err)
    |             --------- ^^^ expected associated type, found `&BlockId<_>`
    |             |
    |             arguments to this method are incorrect
--------------------------------------------------------------------------------------
![1691463385285](https://github.com/paritytech/substrate/assets/141697721/546cc0d3-5d29-4cb6-80e4-e4667fcbb80f)

### Steps to reproduce

1、add rpc configuration to the prject correctly
2、build the project
3、errors like this:
--------------------------------------------------------------------------------------
45  |         api.get_value(&at,v).map_err(runtime_error_into_rpc_err)
    |             --------- ^^^ expected associated type, found `&BlockId<_>`
    |             |
    |             arguments to this method are incorrect
--------------------------------------------------------------------------------------

**Comment by ggwpez** [comment]
Did you follow some tutorial or how did you add it to the template?  
Where did you get the code from?

**Comment by Snikko** [comment]
> Did you follow some tutorial or how did you add it to the template? Where did you get the code from?

yeah,I followed  some tutorials to add it to my template,sunch as  https://substrate.stackexchange.com/questions/5554/how-to-add-custom-rpcs  and  https://github.com/AlexD10S/susbtrate-node-template/blob/rpc-custom-methods/pallets/template/rpc/src/lib.rs#L9  and others. I compared the codes among them and found that they did the same thing.But ,when I added the codes in my project, errors occured like that when it is built.So,maybe,I think  wether  the newest codes causing the problem,I am not sure.

**Comment by ggwpez** [comment]
@AlexD10S do you know if the tutorial maybe needs an update?

**Comment by Snikko** [comment]
> @AlexD10S do you know if the tutorial maybe needs an update?
yeah,I know.I download the newest version of substrate-node-template and add the rpc configuration to it following the tutorial,changing the version to the newest in the file Cargo.toml of  rpc.The version is v1.0.0.(branch = "polkadot-v1.0.0")

**Comment by Snikko** [comment]
> @AlexD10S do you know if the tutorial maybe needs an update?

I have found the answer.Thanks.

**Comment by AlexD10S** [comment]
I am going to take a look, probably it needs an update. The tutorial was done with `polkadot-v0.9.28`. 

@Snikko If you have found an answer, can you share your solution here or in the [StackExchange question](https://substrate.stackexchange.com/questions/5554/how-to-add-custom-rpcs) with a new answer or editing mine. The community will benefit if you share your solution there.

**Comment by Snikko** [comment]
> I am going to take a look, probably it needs an update. The tutorial was done with `polkadot-v0.9.28`.
> 
> @Snikko If you have found an answer, can you share your solution here or in the [StackExchange question](https://substrate.stackexchange.com/questions/5554/how-to-add-custom-rpcs) with a new answer or editing mine. The community will benefit if you share your solution there.

In recent versions,we have to change some codes in the file ***pallets/template/src/rpc/src/lib.rs***,as follows:
--------------------------------------------------------------------
 impl<C, Block> TemplateApiServer<<Block as BlockT>::Hash> for TemplatePallet<C, Block>
where
    Block: BlockT,
    C: Send + Sync + 'static + ProvideRuntimeApi<Block> + HeaderBackend<Block>,
    C::Api: TemplateRuntimeApi<Block>,
{
    fn get_value(&self, at: Option<<Block as BlockT>::Hash>) -> RpcResult<u32> 
    
    {
        let api = self.client.runtime_api();
        let at = at.unwrap_or_else(||self.client.info().best_hash);

        api.get_value(at).map_err(runtime_error_into_rpc_err)
    }
}
--------------------------------------------------------------------
